File: #0135.candy.cpp
class Solution {
public:
    int candy(vector<int>& ratings) {

        int size = ratings.size();
        vector<int> result(size, 1);
        for (int i = 1; i < size; ++i) {
            if (ratings[i] > ratings[i - 1]) {
                result[i] = result[i - 1] + 1;
            }
        }

        int start =  size - 2;
        for (int i = start; i >= 0; --i) {
            if (ratings[i] > ratings[i + 1]) {
                result[i] = std::max(result[i], result[i + 1] + 1);
            }
        }
        return std::accumulate(std::begin(result), std::end(result), 0);
    }
};

File: #0135.candy.js
/**
 * @param {number[]} ratings
 * @return {number}
 */
var candy = function(ratings) {
    const results = new Array(ratings.length).fill(1);
    for (let i = 1; i < ratings.length; i++) {
        if (ratings[i] > ratings[i-1]) {
            results[i] = results[i-1] + 1;
        }
    }

    for (let i = ratings.length - 2; i >= 0; i--) {
        if (ratings[i] > ratings[i+1]) {
            results[i] = Math.max(results[i],
                results[i+1] + 1);
        }
    }
    return results.reduce((left, right) => left + right);
};
